platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/heartbeats/HeartbeatProtocol.java

```diff
@@ -0,0 +1,147 @@
+     * @param heartbeatPeriod The period (milliseconds) at which this protocol should execute
+     * @param networkMetrics  Network metrics, for recording roundtrip heartbeat time
+     */
+    public HeartbeatProtocol(
+            final @NonNull NodeId otherId,
+            final @NonNull Duration heartbeatPeriod,
+            final @NonNull NetworkMetrics networkMetrics) {
+
+        this.otherId = ArgumentUtils.throwArgNull(otherId, "otherId");
+        this.heartbeatPeriod = ArgumentUtils.throwArgNull(heartbeatPeriod, "heartbeatPeriod");
+        this.networkMetrics = ArgumentUtils.throwArgNull(networkMetrics, "networkMetrics");
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Returns true if the last heartbeat protocol was started more than {@link #heartbeatPeriod} ago
+     */
+    @Override
+    public boolean shouldInitiate() {
+        return Duration.between(lastHeartbeatTime, Instant.now()).compareTo(heartbeatPeriod) >= 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldAccept() {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean acceptOnSimultaneousInitiate() {
+        return true;
+    }
+
+    /**
+     * Reads a byte from the socket
+     *
+     * @param connection the connection to read from
+     * @param startTime  the start time measurement for determining roundtrip time
+     * @throws IOException              if there is an error reading from the socket
+     * @throws NetworkProtocolException if the byte read is not a heartbeat or heartbeat ack
+     */
+    private void readByte(final @NonNull Connection connection, final long startTime)
+            throws IOException, NetworkProtocolException {
+
+        final byte readByte = connection.getDis().readByte();
+
+        if (readByte == ByteConstants.HEARTBEAT) {
+            // if we received their heartbeat, immediately send an ack
+            connection.getDos().write(ByteConstants.HEARTBEAT_ACK);
+            connection.getDos().flush();
+        } else if (readByte == ByteConstants.HEARTBEAT_ACK) {
+            // the time from the measurement start to now represents the total roundtrip
+            networkMetrics.recordPingTime(otherId, System.nanoTime() - startTime);
+        } else {
+            throw new NetworkProtocolException(String.format(
+                    "received %02x but expected %02x or %02x (HEARTBEAT or HEARTBEAT_ACK)",
+                    readByte, ByteConstants.HEARTBEAT, ByteConstants.HEARTBEAT_ACK));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void runProtocol(final @NonNull Connection connection)
+            throws NetworkProtocolException, IOException, InterruptedException {
+
+        // record the time prior to executing the protocol, so that heartbeatPeriod represents a true period, as
+        // opposed to a sleep time
+        lastHeartbeatTime = Instant.now();
```

Comment:
I'd suggest using a `Time` object here to get the current time (which you'd have to pass into the constructor), instead of using `Instant.now()`.  This line would become

```suggestion
        // record the time prior to executing the protocol, so that heartbeatPeriod represents a true period, as
        // opposed to a sleep time
        lastHeartbeatTime = time.now();
```

`Time` is a utility class we created, you can find it in `com.swirlds.common.time`. The reason we use this level of indirection is because it allows us to spoof the clock during unit tests. If you wanted to test that this class only accepted after a certain time period elapses, you can simulate that with the `FakeTime` implementation. In production environments, this class will be provided with an implementation that provides the actual wall clock time.

Reply:
done